main.go

```diff
@@ -16,33 +16,269 @@ limitations under the License.
+	flag.Parse()
+	if *outputPtr == "" {
+		*outputPtr = *algPtr + "_result.csv"
+	}
 
-	alg, err := modeling.CreateAlg(0.4, 100)
+	var errList []<-chan error
+
+	zoneNames, inputQueue, errC, err := parseInput(*inputPtr)
 	if err != nil {
 		panic(err)
 	}
-
-	var sim modeling.TheoreticalSimulator
-
-	// TODO: operations on metrics rather than printing only
-	model, err := modeling.NewModel(zones, alg, sim)
+	errList = append(errList, errC)
+	outputQueue, errC, err := startSimulation(*algPtr, inputQueue)
 	if err != nil {
 		panic(err)
 	}
-
-	err = model.StartSimulation()
+	errList = append(errList, errC)
+	errC, err = parseResult(*outputPtr, zoneNames, outputQueue)
 	if err != nil {
 		panic(err)
 	}
-	err = model.PrintLastResult()
+	errList = append(errList, errC)
+	// wait for goroutings to return
+	err = waitForSimulation(errList...)
 	if err != nil {
 		panic(err)
 	}
 }
+
+// every row of the input file will be parsed to one instance of inputData
+type inputData struct {
+	// input id of the row
+	id string
+	// parse zone info of the input file into zone data structure
+	zones []modeling.Zone
+}
+
+// every piece of input data will be processed to one instance of outputData
+type outputData struct {
+	// same id as input id
+	id string
+	// simulation result of that piece of input data
+	result modeling.SimulationResult
+}
+
+// paserInput parses input csv file to instances of inputData and puts them into
+// a queue
+func parseInput(file string) ([]string, <-chan inputData, <-chan error, error) {
+	inputFile, err := os.Open(filepath.Join("", filepath.Clean(file)))
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	fmt.Printf("Reading data from %v\n", file)
+	reader := csv.NewReader(inputFile)
+	reader.TrimLeadingSpace = true
+	line, err := reader.Read()
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	var zoneNames []string
+	for _, name := range line[1:] {
+		name = strings.TrimSpace(name)
+		zoneNames = append(zoneNames, name)
+	}
+	dataC := make(chan inputData)
+	errC := make(chan error, 1)
+
+	go func() {
+		defer close(dataC)
+		defer close(errC)
+		for {
+			// read one row
+			datas, err := reader.Read()
+			if err == io.EOF {
+				break
+			}
+			if err != nil {
+				errC <- err
+				return
+			}
+			if len(datas) != len(zoneNames)+1 {
+				fmt.Printf("[WARNING] unmatched number of fields, skip data %v\n", datas)
+				continue
+			}
+			var oneData inputData
+			oneData.id = datas[0]
+			for index, data := range datas[1:] {
+				nodestr := strings.Fields(data)
+				// convert string to int. number of nodes in a zone
+				numNode, err := strconv.Atoi(nodestr[0])
+				if err != nil {
+					errC <- err
+					return
+				}
+				// convert string to int. number of endpoints in a zone
+				numEndpoints, err := strconv.Atoi(nodestr[1])
+				if err != nil {
+					errC <- err
+					return
+				}
+				oneData.zones = append(oneData.zones, modeling.Zone{Nodes: numNode, Endpoints: numEndpoints, Name: zoneNames[index]})
```

Comment:
This line feels a bit long, maybe this would be better?
```suggestion
				oneData.zones = append(oneData.zones, modeling.Zone{
					Nodes: numNode,
					Endpoints: numEndpoints,
					Name: zoneNames[index],
				})
```

Reply:
changed